projects/nginx/fuzz/http_request_fuzzer.c

```diff
@@ -0,0 +1,256 @@
+"        proxy_set_header X-Real-IP $remote_addr;\n"
+"        proxy_set_header X-Real-Port $remote_port;\n"
+"        proxy_set_header Connection '';\n"
+"        chunked_transfer_encoding off;\n"
+"        proxy_buffering off;\n"
+"        proxy_cache off;\n"
+"      }\n"
+"    }\n"
+"}\n"
+"\n";
+
+ssize_t __wrap_listen(int fd, void* buf, size_t bytes) {
+  return 0;
+}
+
+static ngx_cycle_t *cycle;
+static ngx_log_t ngx_log;
+static ngx_open_file_t ngx_log_file;
+static char *my_argv[2];
+static char arg1[] = { 0, 0xAA, 0 };
+
+extern char **environ;
+
+static char *socket_name = "nginx.sock";
+static char *config_file = "socket_config.conf";
+
+// Create a base state for Nginx without starting the server
+char *InitializeNginx(void) {
+  ngx_log_t *log;
+  ngx_cycle_t init_cycle;
+
+  ngx_debug_init();
+  ngx_strerror_init();
+  ngx_time_init();
+#if (NGX_PCRE)
+  ngx_regex_init();
+#endif
+
+  // Just output logs to stderr
+  ngx_log.file = &ngx_log_file;
+  ngx_log.log_level = NGX_LOG_EMERG;
+  ngx_log_file.fd = ngx_stderr;
+  log = &ngx_log;
+
+#if (NGX_OPENSSL)
+  ngx_ssl_init(log);
+#endif
+  ngx_memzero(&init_cycle, sizeof(ngx_cycle_t));
+  init_cycle.log = log;
+  ngx_cycle = &init_cycle;
+
+  init_cycle.pool = ngx_create_pool(1024, log);
+
+  // Set custom argv/argc
+  my_argv[0] = arg1;
+  my_argv[1] = NULL;
+  ngx_argv = ngx_os_argv = my_argv;
+  ngx_argc = 0;
+
+  // Weird trick to free a leaking buffer always caught by ASAN
+  // We basically let ngx overwrite the environment variable, free the leak and
+  // restore the environment as before.
+  char *env_before = environ[0];
+  environ[0] = my_argv[0] + 1;
+  ngx_os_init(log);
+  free(environ[0]);
+  environ[0] = env_before;
+
+  ngx_crc32_table_init();
+  ngx_preinit_modules();
+
+  FILE *fptr = fopen("socket_config.conf", "w");
+  fprintf(fptr, configuration);
+  fclose(fptr);
+  init_cycle.conf_file.len = strlen(config_file);
+  init_cycle.conf_file.data = (unsigned char *) config_file;
+
+  cycle = ngx_init_cycle(&init_cycle);
+
+  ngx_os_status(cycle->log);
+  ngx_cycle = cycle;
+  return config_file;
+}
+
+// If this function is called, the fuzzer needs to be updated
+void invalid_call(void) { }
+
+struct fuzzing_data {
+  const uint8_t *data;
+  size_t data_len;
+};
+
+// Called by the http parser to read the buffer
+static ssize_t recv_handler(ngx_connection_t *c, u_char *buf, size_t size) {
+  struct fuzzing_data *data = (struct fuzzing_data*)(c->write->data);
+  if (data->data_len < size)
+    size = data->data_len;
+  memcpy(buf, data->data, size);
+  data->data += size;
+  data->data_len -= size;
+  return size;
+}
+
+
+// Used when sending data, basically do nothing
+ngx_chain_t *send_chain(ngx_connection_t *c, ngx_chain_t *in, off_t limit) {
+  return in;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t data_len) {
+  if( access("nginx.sock", F_OK ) != -1 ) {
+    remove("nginx.sock");
+  }
+
+  InitializeNginx();
```

Comment:
does it have to be called on every execution? the best practice is to do it only once, see http://llvm.org/docs/LibFuzzer.html#startup-initialization